### PR TITLE
Add bikes app (VéloStar and Vélib')

### PR DESCRIPTION
**Pull request by vrialland**
A review on this is welcome as it is my first Django app ;)

### CODE REVIEW

File: app_bikes/settings.py [hunk 1]
```diff
@@ -0,0 +1,3 @@
+VALUES_UPDATE_INTERVAL = 10
+STAR_API_KEY = ''
+VELIB_API_KEY = ''
```
Comment (magopian) [inline app_bikes/settings.py]: Why have those here if they're empty? Did you mean to have the base urls in here instead of in the provider modules?

**Comment by magopian** [comment]
Oh and btw, I filed #26: when an app is first "configured" for a box, it should update straight away, and not for the first time only after the delay.

I'm going to try fixing that for the existing/merged apps.

**Comment by magopian** [comment]
Well, it seems it's only in this app (and mine which isn't merged yet) that we have the issue: we need to check if the delay is elapsed OR if the fields aren't filled in. I'm going to fix that in my PR, maybe you should fix it here too ;)

File: app_bikes/apps.py [hunk 2]
```diff
@@ -0,0 +1,11 @@
+# coding: UTF-8
+
+from __future__ import unicode_literals
+
+from django.apps import AppConfig
+from django.utils.translation import ugettext_lazy as _L
```
Comment (vrialland) [inline app_bikes/apps.py]: some old habits from another framework :) _ -> ugettext, _L -> ugettext_lazy, just to know which one you're using

File: app_bikes/admin.py [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+from .models import AppBikes
+from .forms import BikeModelForm
+
+
+class AppBikesAdmin(admin.ModelAdmin):
+    readonly_fields = ('created_date', 'last_activity', 'slots', 'bikes', 'status')
+
+    list_display = ('boite', 'enabled', 'get_app_dictionary', 'created_date', 'last_activity')
+
+    exclude = ('station',)
```
Comment (vrialland) [inline app_bikes/admin.py]: It's to avoid it being displayed twice in station and in station_id autocomplete

File: app_bikes/forms.py [hunk 4]
```diff
@@ -0,0 +1,26 @@
+from dal import autocomplete
+from django import forms
+
+
+class BikeModelForm(forms.ModelForm):
+    def __init__(self, *args, **kw):
+        super(BikeModelForm, self).__init__(*args, **kw)
+        if self.instance is not None:
+            # Saved instance is loaded, setup choices to display the selected value
+            self.fields['id_station'].widget.choices = ((self.instance.id_station, self.instance.station),)
```
Comment (vrialland) [inline app_bikes/forms.py]: This was the tricky part and I'm not very statisfied of it. I chose to store id_station, which will be the invariant identifier of the station, and its name (station).
The station field has no other purpose than to be in the JSON output if we want to display the station (in case we have several).
This field is base on a SelectField, which has a predefined list of choices. I could have called the webservice to get all the (id, name) pairs for all the stations of all the providers but it's very time consuming, so to avoid that I have to overwrite the choices:
- when the user inputs something (the autocomplete will be triggered)
- when the user wants to edit the model (this case), I rebuild the choices with only one value, so the autocomplete will display the current station (that's also why "station" field is not displayed in the form)
- In case there was an error in the form, to set its value back to what was selected.

tldr; : it's the second case in your question, hope my answer was clear

File: app_bikes/forms.py [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+
+    def validate_unique(self):
```
Comment (vrialland) [inline app_bikes/forms.py]: Done

File: app_bikes/providers/__init__.py [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+    'velib': VelibProvider
+}
+
+
+def get_provider(key):
+    return PROVIDERS.get(key)
+
+
+def get_providers():
+    ret = [(key, cls.verbose_name) for key, cls in PROVIDERS.iteritems()]
```
Comment (vrialland) [inline app_bikes/providers/__init__.py]: done

File: app_bikes/settings.py (hunk 1 above)
Comment (vrialland) [inline app_bikes/settings.py]: Empty keys work when developing, not sure we should store api_keys as they are personal data.

File: app_bikes/providers/star.py [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+    STAR_API_BASE_URL = 'https://data.explore.star.fr/api/records/1.0/search'
+
+    verbose_name = _L(u'Star - Rennes')
+
+    @classmethod
+    def get_stations(cls, query):
+        if query.isdigit():
+            query = 'nom:{query} OR idstation:{query}'.format(query=query)
+        else:
+            # As idstation is an int in the API it raises an error otherwise
```
Comment (vrialland) [inline app_bikes/providers/star.py]: if the query is a string, the call will raise an error because idstation is store as an int in the webservice database, so we have to remove this field

**Comment by vrialland** [comment]
- chmod on manage.py was not intended: I suspect GIT on Windows to have done that.
- No problem with Firefox 45.0.1, what is your version?

File: app_bikes/forms.py (hunk 4 above)
Comment (magopian) [inline app_bikes/forms.py]: ok, I think I got it. I haven't checked into the autocompletion library code/doc, so dunno if there's an easier way to do that. It does seem a bit hacky though (and doesn't seem to work for me at least, as I commented, editing a configuration doesn't pre-fill the autocomplete field).

**Comment by magopian** [comment]
My version is Firefox Beta (50.0b8).

**Comment by magopian** [comment]
The PR has been merged, so you won't be able to "update" it by pushing commits on the branch as you used to do before it was merged. You'll have to create a new PR.

**Comment by vrialland** [comment]
New PR https://github.com/bgaultier/laboitepro/pull/28
